### PR TITLE
feat: add openai tool parameter

### PR DESCRIPTION
**Pull request by Undertone0809**
Add openai tool parameter
https://platform.openai.com/docs/api-reference/chat/create#chat-create-tools

### CODE REVIEW

File: src/providers/openai.ts [hunk 1]
```diff
@@ -30,6 +30,8 @@ type OpenAiCompletionOptions = OpenAiSharedOptions & {
   best_of?: number;
   functions?: OpenAiFunction[];
   function_call?: 'none' | 'auto' | { name: string };
+  tools?: OpenAiTool[];
+  tool_choice?: 'none' | 'auto' | { name: string };
```
Comment (typpo) [inline src/providers/openai.ts]: According to the [OpenAI docs](https://platform.openai.com/docs/api-reference/chat/create#chat-create-tools), should this be
```ts
{
  // ...
  tools_choice?: 'none' | 'auto' | { type: 'function'; function?: { name: string }}
}
```

File: src/providers/openaiUtil.ts [hunk 2]
```diff
@@ -8,6 +8,11 @@ export interface OpenAiFunction {
   parameters: any;
 }
 
+export interface OpenAiTool {
+  type: string;
+  function: any;
```
Comment (typpo) [inline src/providers/openaiUtil.ts]: Maybe we can type this as 
```ts
export interface OpenAiTool {
  type: string;
  function: OpenAiFunction;
}
```

**Review by typpo — CHANGES_REQUESTED** [review]
This is a good start!  You'll also have to add `tools` and `tool_choice` to the request to OpenAI in `OpenAiChatCompletionProvider.callApi`.

File: src/providers/openai.ts (hunk 1 above)
Comment (Undertone0809) [reply src/providers/openai.ts]: done

File: src/providers/openaiUtil.ts (hunk 2 above)
Comment (Undertone0809) [reply src/providers/openaiUtil.ts]: done

**Review by typpo — APPROVED** [review]
Thanks for the change!  I also added docs and modified the `callApi` function to pass the `tools` and `tool_choice` params

**Comment by Undertone0809** [comment]
@typpo Next stage, I think we can add relevant assert.

**Comment by typpo** [comment]
@Undertone0809 `is-valid-openai-tools-call` assert type added in #354

**Comment by typpo** [comment]
@Undertone0809 `is-valid-openai-tools-call` assert type added in #354

**Pull request merged**